### PR TITLE
feat: get users' notifications

### PR DESCRIPTION
**Pull request by zee-ham-su**
## What does the PR do?
This PR addresses issue of implementing the functionality to retrieve notifications for authenticated users from the database.  The endpoint returns errors if:

- [x] User is not logged in.
- [x] User does not exist.
- [x] Validation errors on the user input is done

## How should this be manually tested?
1. Login to a user account, and get the access token
2. You can send a GET request to the following endpoint:
GET /api/v1/notifications

> ```json
> Authorization: Bearer <access_token>
> ```

### Success Response
on a successful execution, you will receive the following response structure:
```
{
    "status": "success",
    "status_code": 200,
    "message": "Notifications retrieved successfully",
    "data": {
        "total_notification_count": 2,
        "total_unread_notification_count": 2,
        "notifications": [
            {
                "notification_id": "0bad0dae-a11a-41e4-beb7-230808c1f3d5",
                "is_read": false,
                "message": "Notification 1 for John",
                "created_at": "2024-07-28T19:29:13.049Z"
            },
            {
                "notification_id": "f16e9a12-b428-4bcc-b9d9-e560758e15c8",
                "is_read": false,
                "message": "Notification 2 for John",
                "created_at": "2024-07-28T19:29:13.049Z"
            }
        ]
    }
}
```
### "Invalid Input" Response
If the request is invalid, you will receive the following response:
```
{
  "status_code": 400,
  "message": "Invalid request.",
  "data": null
}

```
### Unauthorized Response
```
{
  "status_code": 401,
  "status": "unauthorized",
  "message": "Unauthorized. Please log in."
}
```

### "Internal Server Error" Response

```
{
  "status_code": 500,
  "status": "Internal server error",
  "message": "Failed to retrieve notifications.",
  "data": null
}

```

## Task
- [x]  Implemented GET /api/v1/notifications endpoint in notifications.controller.ts
- [x] Added method to fetch notifications in the NotificationsService.
- [x] Tested endpoint functionality to ensure it handles various scenarios correctly
- [x] Added appropriate HTTP status codes and response messages for edge cases. 
- [x] Added unit tests for the NotificationsService.
- [x] Updated API documentation to reflect the new endpoint

Reference Issue: [Link to Issue](https://linear.app/team-bingo/issue/BAC2-42/retrieve-all-notifications-for-a-user)

### CODE REVIEW

File: src/modules/notifications/notifications.controller.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
+import { Controller, Get, Req } from '@nestjs/common';
+import { NotificationsService } from './notifications.service';
+import { UserPayload } from '../user/interfaces/user-payload.interface';
+import { ApiBearerAuth, ApiTags, ApiResponse } from '@nestjs/swagger';
+
+@ApiBearerAuth()
+@ApiTags('notifications')
```
Comment (CFCIfe) [inline src/modules/notifications/notifications.controller.ts:7]:
```suggestion
@ApiTags('Notifications')
```

File: src/modules/notifications/notifications.controller.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+            {
+              notification_id: 2,
+              is_read: true,
+              message: 'Your order has been shipped',
+              created_at: '2023-07-29T11:22:33Z',
+            },
+          ],
+        },
+      },
+    },
```
Comment (CFCIfe) [inline src/modules/notifications/notifications.controller.ts:40]: Just add a type of your dto

For example.
```suggestion
type: NotificationResponseSuccessDto
```

File: src/modules/notifications/notifications.controller.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  @ApiResponse({
+    status: 500,
+    description: 'Failed to retrieve notifications.',
+    schema: {
+      example: {
+        status: 'error',
+        message: 'Failed to retrieve notifications.',
+        status_code: 500,
+        data: null,
+      },
```
Comment (CFCIfe) [inline src/modules/notifications/notifications.controller.ts:51]: Use type of your DTO

File: src/modules/notifications/notifications.controller.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+              is_read: true,
+              message: 'Your order has been shipped',
+              created_at: '2023-07-29T11:22:33Z',
+            },
+          ],
+        },
+      },
+    },
+  })
+  @ApiResponse({
```
[the remark this answers is not in the document]
Comment (CFCIfe) [reply src/modules/notifications/notifications.controller.ts:42]: Add more for:
@ApiBadRequestResponse()
@ApiUnauthorizedResponse()

File: src/modules/notifications/notifications.controller.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        data: {
+          total_notification_count: notifications.totalNotificationCount,
+          total_unread_notification_count: notifications.totalUnreadNotificationCount,
+          notifications: notifications.notifications.map(notification => ({
+            notification_id: notification.id,
+            is_read: notification.is_Read,
+            message: notification.message,
+            created_at: notification.created_at,
+          })),
+        },
```
Comment (CFCIfe) [inline src/modules/notifications/notifications.controller.ts:72]:
```suggestion
 data: {
      total_notification_count: totalNotificationCount,
      total_unread_notification_count: totalUnreadNotificationCount,
      notifications: notifications.map(({ id, is_Read, message, created_at }) => ({
        notification_id: id,
        is_read: is_Read,
        message,
        created_at,
      })),
    },
```

File: src/modules/notifications/notifications.controller.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+      };
+    } catch (error) {
+      return {
+        status: 'error',
+        message: 'Failed to retrieve notifications.',
+        status_code: 500,
+        data: null,
```
Comment (CFCIfe) [inline src/modules/notifications/notifications.controller.ts:79]: We agreed to remove this. Please take it off

File: src/modules/notifications/notifications.controller.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+      },
+    },
+  })
+  async getNotifications(@Req() req: { user: UserPayload }) {
+    const userId = req.user.id;
+    try {
+      const notifications = await this.notificationsService.getNotificationsForUser(userId);
+
```
[the remark this answers is not in the document]
Comment (zee-ham-su) [reply src/modules/notifications/notifications.controller.ts:58]: it is checked in the service

File: src/modules/notifications/notifications.controller.ts (hunk 5 above)
Comment (zee-ham-su) [reply src/modules/notifications/notifications.controller.ts:72]: comment resolved

File: src/modules/notifications/notifications.controller.ts (hunk 3 above)
Comment (zee-ham-su) [reply src/modules/notifications/notifications.controller.ts:51]: resolved

File: src/modules/notifications/notifications.controller.ts (hunk 2 above)
Comment (zee-ham-su) [reply src/modules/notifications/notifications.controller.ts:40]: resolved

File: src/modules/notifications/notifications.controller.ts (hunk 3 above)
Comment (zee-ham-su) [reply src/modules/notifications/notifications.controller.ts:51]: resolved

File: src/modules/notifications/dtos/get-notification-dto.ts
Comment (CFCIfe) [inline src/modules/notifications/dtos/get-notification-dto.ts:1]: The filename and DTO class have 2 different names. Fix please

File: src/modules/organisations/dto/organisation.dto.ts
Comment (CFCIfe) [inline src/modules/organisations/dto/organisation.dto.ts:1]: REVERT!!!! Not your file.

File: src/modules/organisations/entities/org-members.entity.ts [hunk 8]
```diff
@@ -12,7 +12,7 @@ export class OrganisationMember extends AbstractBaseEntity {
   @ManyToOne(() => Organisation, organisation => organisation.organisationMembers)
   organisation_id: Organisation;
 
-  @Column({ nullable: true })
+  @Column()
```
Comment (CFCIfe) [inline src/modules/organisations/entities/org-members.entity.ts:15]: Revert back to

```suugestion
@Column({ nullable: true })
```

File: src/modules/organisations/organisations.controller.ts [hunk 9]
```diff
@@ -27,15 +27,6 @@ import { OrganisationMembersResponseDto } from './dto/org-members-response.dto';
...
-  @ApiOperation({ summary: 'Create new Organisation' })
-  @ApiResponse({
-    status: 201,
-    description: 'The created organisation',
-  })
-  @ApiResponse({
-    status: 409,
-    description: 'Organisation email already exists',
-  })
   @Post('/')
```
Comment (CFCIfe) [inline src/modules/organisations/organisations.controller.ts:30]:
```suggestion
 @ApiOperation({ summary: 'Create new Organisation' })
  @ApiResponse({
    status: 201,
    description: 'The created organisation',
  })
  @ApiResponse({
    status: 409,
    description: 'Organisation email already exists',
  })
```

File: src/modules/organisations/organisations.module.ts [hunk 10]
```diff
@@ -5,10 +5,9 @@ import { TypeOrmModule } from '@nestjs/typeorm';
 import { Organisation } from './entities/organisations.entity';
 import { User } from '../user/entities/user.entity';
 import { UserModule } from '../user/user.module';
-import { OrganisationMember } from './entities/org-members.entity';
 
 @Module({
-  imports: [TypeOrmModule.forFeature([Organisation, User, OrganisationMember]), UserModule],
+  imports: [TypeOrmModule.forFeature([Organisation, User]), UserModule],
```
Comment (CFCIfe) [inline src/modules/organisations/organisations.module.ts:10]:
```suggestion
  imports: [TypeOrmModule.forFeature([Organisation, User, OrganisationMember]), UserModule],
```

File: src/modules/organisations/organisations.module.ts [hunk 11]
```diff
@@ -5,10 +5,9 @@ import { TypeOrmModule } from '@nestjs/typeorm';
 import { Organisation } from './entities/organisations.entity';
 import { User } from '../user/entities/user.entity';
 import { UserModule } from '../user/user.module';
```
Comment (CFCIfe) [inline src/modules/organisations/organisations.module.ts:7]: Revert the import of `OrganisationMember`

File: src/modules/organisations/organisations.service.ts [hunk 12]
```diff
@@ -21,17 +19,14 @@ import { CreateOrganisationMapper } from './mapper/create-organisation.mapper';
...
 
 @Injectable()
 export class OrganisationsService {
   constructor(
     @InjectRepository(Organisation)
     private readonly organisationRepository: Repository<Organisation>,
     @InjectRepository(User)
-    private readonly userRepository: Repository<User>,
-    @InjectRepository(OrganisationMember)
-    private readonly organisationMemberRepository: Repository<OrganisationMember>
```
Comment (CFCIfe) [inline src/modules/organisations/organisations.service.ts:34]: Please revert this back too

```suggestion
    @InjectRepository(OrganisationMember)
    private readonly organisationMemberRepository: Repository<OrganisationMember>
```

**Review by CFCIfe — COMMENTED** [review]
Checkout all the following files on DEV, and revert all the changes you made on them (unknowingly). Just copy and paste what is on dev in them (Make sure to sync branch before doing it).

```
src/modules/organisations/dto/organisation.dto.ts
src/modules/organisations/entities/org-members.entity.ts
src/modules/organisations/organisations.controller.ts
src/modules/organisations/organisations.module.ts
src/modules/organisations/organisations.service.ts
src/modules/organisations/tests/organisations.service.spec.ts
src/modules/waitlist/waitlist.controller.ts
```

**Review by CFCIfe — COMMENTED** [review]
Satisfied. Request approval.

**Review by rugue — APPROVED** [review]

**Review by davidolanrewaju — APPROVED** [review]

**Review by davidolanrewaju — DISMISSED** [review]

**Review by TheCodeGhinux — APPROVED** [review]
Okayy

**Review by Cobidex — APPROVED** [review]

**Review by Heba-WebDev — APPROVED** [review]

**Review by Homoakin619 — APPROVED** [review]

**Review by davidolanrewaju — APPROVED** [review]
